GVRf/Extensions/ResonanceAudio/resonanceaudio/src/main/java/org/gearvrf/resonanceaudio/GVRAudioListener.java

```diff
@@ -0,0 +1,294 @@
+            throw new UnsupportedOperationException("Audio source must be at the same root as audio listener");
+        }
+
+        updateSoundPosition(audioSource);
+        mAudioEngine.resumeSound(audioSource.getSourceId());
+    }
+
+    void stopSound(GVRAudioSource audioSource) {
+        mAudioEngine.stopSound(audioSource.getSourceId());
+    }
+
+    void setSoundVolume(GVRAudioSource audioSource, float volume) {
+        mAudioEngine.setSoundVolume(audioSource.getSourceId(), volume);
+    }
+
+    boolean isSoundPlaying(GVRAudioSource audioSource) {
+        if (audioSource.getSourceId() == GvrAudioEngine.INVALID_ID) {
+            return false;
+        }
+
+        return mAudioEngine.isSoundPlaying(audioSource.getSourceId());
+    }
+
+    private void updateSoundPosition(GVRAudioSource audioSource) {
+        GVRTransform t = audioSource.getTransform();
+
+        //TODO: Should move it to audio thread!?
+        mAudioEngine.setSoundObjectPosition(audioSource.getSourceId(),
+                t.getPositionX(), t.getPositionY(), t.getPositionZ());
+    }
+
+    private void updateHeadTransform() {
+        GVRTransform owner = mScene.getMainCameraRig().getOwnerObject().getTransform();
+        Quaternionf rotation = new Quaternionf(owner.getRotationX(),
+                owner.getRotationY(), owner.getRotationZ(), owner.getRotationW());
+
+        GVRTransform head = mScene.getMainCameraRig().getHeadTransform();
+        rotation.mul(head.getRotationX(),
+                head.getRotationY(), head.getRotationZ(), head.getRotationW());
+
+        mAudioEngine.setHeadPosition(owner.getPositionX(),
+                owner.getPositionY(), owner.getPositionZ());
+        mAudioEngine.setHeadRotation(rotation.x, rotation.y, rotation.z, rotation.w);
+    }
+
+    private void update() {
+        if (!isEnabled()) {
+            return;
+        }
+
+        updateHeadTransform();
+
+        for (GVRAudioSource audioSource: mAudioSources) {
+            updateSoundPosition(audioSource);
+        }
+
+        // TODO: maybe the following method should not be called from rendering thread
+        mAudioEngine.update();
+    }
+
+    @Override
+    public void setOwnerObject(GVRSceneObject owner) {
+        // This component will only be attached to scene root object
+        if (owner != null && owner != mScene.getRoot()) {
+            throw new UnsupportedOperationException("Audio listener can not change owner");
+        }
+
+        super.setOwnerObject(owner);
+    }
+
+    private static class AudioEngineEventsHandler extends GVREventListeners.ActivityEvents
+            implements GVRDrawFrameListener {
+        private static WeakReference<AudioEngineEventsHandler> mInstance = new WeakReference<AudioEngineEventsHandler>(null);
+
+        private GVRContext mContext;
+        private GVRAudioListener mCurrentListener;
+        private GVRScene mCurrentScene;
+
+        public static synchronized AudioEngineEventsHandler getInstance() {
```

Comment:
Hi @liaxim,
we have considered that issues during our development but we didn't figure out a way to get notifications when the main scene is changing.

When we change from SceneA to SceneB we need pause the audio sources from SceneA and resume the audio sources from SceneB. But to do that we need listen to main scene's changes.

Because of that we implemented the singleton. :(

But I agreen that we need a beter logic to fix that issue. Do you have any suggestion?

Reply:
@liaxim You can see at line 245 the code to check scene's changes.